presto-main/src/main/java/io/prestosql/execution/buffer/PagesSerde.java

```diff
@@ -56,86 +56,225 @@ public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> co
+        //  based on length so that they can be used for compression or encryption, maximizing reuse opportunities
+        private byte[] largerBuffer;
+        private byte[] smallerBuffer;
+        private boolean closed;
+
+        private void checkNotClosed()
+        {
+            if (closed) {
+                throw new IllegalStateException("PagesSerdeContext is already closed");
+            }
+        }
 
-            slice = Slices.wrappedBuffer(decrypted, 0, decryptedSize);
+        private DynamicSliceOutput acquireSliceOutput(int estimatedSize)
+        {
+            checkNotClosed();
+            if (sliceOutput != null && sliceOutput.writableBytes() >= estimatedSize) {
+                DynamicSliceOutput result = this.sliceOutput;
+                this.sliceOutput = null;
+                return result;
+            }
+            this.sliceOutput = null; // Clear any existing slice output that might be smaller than the request
+            return new DynamicSliceOutput(estimatedSize);
         }
 
-        if (serializedPage.isCompressed()) {
-            checkState(decompressor.isPresent(), "Page is compressed, but decompressor is missing");
+        private void releaseSliceOutput(DynamicSliceOutput sliceOutput)
+        {
+            if (closed) {
+                return;
+            }
+            sliceOutput.reset();
+            if (sliceOutput.writableBytes() <= MAX_BUFFER_RETAINED_SIZE) {
+                this.sliceOutput = sliceOutput;
+            }
+        }
 
-            int uncompressedSize = serializedPage.getUncompressedSizeInBytes();
-            byte[] decompressed = new byte[uncompressedSize];
-            checkState(decompressor.get().decompress(
-                    slice.byteArray(),
-                    slice.byteArrayOffset(),
-                    slice.length(),
-                    decompressed,
-                    0,
-                    uncompressedSize) == uncompressedSize);
+        private byte[] acquireBuffer(int size)
+        {
+            checkNotClosed();
+            byte[] result;
+            //  Check the smallest buffer first
+            if (smallerBuffer != null && smallerBuffer.length >= size) {
+                result = smallerBuffer;
+                smallerBuffer = null;
+                return result;
+            }
+            if (largerBuffer != null && largerBuffer.length >= size) {
+                result = largerBuffer;
+                largerBuffer = smallerBuffer;
+                smallerBuffer = null;
+                return result;
+            }
+            return new byte[size];
+        }
 
-            slice = Slices.wrappedBuffer(decompressed);
+        private void releaseBuffer(byte[] buffer)
+        {
+            if (closed || buffer.length > MAX_BUFFER_RETAINED_SIZE) {
+                return;
+            }
+            if (largerBuffer == null) {
+                largerBuffer = buffer;
+            }
+            else if (smallerBuffer == null) {
```

Comment:
The first time a buffer is released, `largerBuffer` will be set unconditionally. The second time it's released, `smallerBuffer` will be set unconditionally. If the second buffer is larger than the first one, we end up with the variables tracking the opposite buffers they should be tracking:

```java
PagesSerdeContext context = new PagesSerdeContext();

context.releaseBuffer(context.acquireBuffer(5));
context.releaseBuffer(context.acquireBuffer(10));

System.out.println("smaller: " + context.smallerBuffer.length);
System.out.println("larger:  " + context.largerBuffer.length);
```

produces:

```
smaller: 10
larger:  5
```

Reply:
Very good catch. This logical bug could only serve to increase the allocation rate slightly so I re-ran the benchmarks after fixing.

Getting this logic correct reduces allocation rate when both compression and encryption are enabled. Here's the previous benchmark output vs the output when fixed: [Bugfix Improvement](https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/pettyjamesm/5177ed85f452df7708e28e15e1f55bfc/raw/d235d264463b9e8b1e77f740e2f1bab9ea177f59/pages-serde-context-benchmark.json,https://gist.githubusercontent.com/pettyjamesm/5177ed85f452df7708e28e15e1f55bfc/raw/2b602642e71575e0a85a271b596b3d2120968f42/pages-serde-context-again.json)

And against baseline: [Updated Benchmark vs Baseline](https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/pettyjamesm/5177ed85f452df7708e28e15e1f55bfc/raw/d235d264463b9e8b1e77f740e2f1bab9ea177f59/baseline-serde-benchmark.json,https://gist.githubusercontent.com/pettyjamesm/5177ed85f452df7708e28e15e1f55bfc/raw/2b602642e71575e0a85a271b596b3d2120968f42/pages-serde-context-again.json)